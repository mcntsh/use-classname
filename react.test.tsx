import React from 'react'
import renderer from 'react-test-renderer'
import { ClassNameCreator } from './'
import withSetClassName from './react'

type ReactComponentProps = {
  type?: 'A' | 'B' | 'C'
  color?: 'B' | 'D'
  id?: string
  children?: React.ReactNode
  setClassName: ClassNameCreator
}

function ReactComponent({ type, children, setClassName, ...props }: ReactComponentProps) {
  const aVariant = {
    when: { type: 'A' },
    use: 'a',
  }
  const abVariant = {
    when: { type: 'A', color: 'B' },
    use: 'ab',
  }

  return (
    <div className={setClassName(aVariant, abVariant)} {...props}>
      {children}
    </div>
  )
}

test('renders the wrapped component as you would expect using the withSetClassName HOC', () => {
  const ReactComponentHOC = withSetClassName(ReactComponent)

  const component = renderer.create(
    <ReactComponentHOC id="foo">
      <div className="child">Hello!</div>
    </ReactComponentHOC>
  )
  const tree = component.toJSON()

  expect(tree).toMatchSnapshot()
})

test('passes the setClassName helper and reacts to prop changes', () => {
  const ReactComponentHOC = withSetClassName(ReactComponent)

  const component = renderer.create(
    <>
      <ReactComponentHOC id="1" type='C' color='D'>
        <div className="child">No classes</div>
      </ReactComponentHOC>
      <ReactComponentHOC id="2" type='A' color='D'>
        <div className="child">Type classes</div>
      </ReactComponentHOC>
      <ReactComponentHOC id="3" type='A' color='B'>
        <div className="child">Type color classes</div>
      </ReactComponentHOC>
    </>
  )
  const tree = component.toJSON()

  expect(tree).toMatchSnapshot()
})

test('passes the debug option to the setClassName helper', () => {
  const ReactComponentHOC = withSetClassName(ReactComponent, { debug: true })

  const component = renderer.create(
      <ReactComponentHOC id="3" type='A' color='B'>
        <div className="child">Type color classes</div>
      </ReactComponentHOC>
  )
  const tree = component.toJSON()

  expect(tree).toMatchSnapshot()
})

test('passes the prefix option to the setClassName helper', () => {
  const ReactComponentHOC = withSetClassName(ReactComponent, { prefix: 'tw-' })

  const component = renderer.create(
      <ReactComponentHOC id="3" type='A' color='B'>
        <div className="child">Type color classes</div>
      </ReactComponentHOC>
  )
  const tree = component.toJSON()

  expect(tree).toMatchSnapshot()
})
