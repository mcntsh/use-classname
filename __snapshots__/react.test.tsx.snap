// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the debug option to the setClassName helper 1`] = `
<div
  className="
• a
• ab"
  color="B"
  id="3"
>
  <div
    className="child"
  >
    Type color classes
  </div>
</div>
`;

exports[`passes the prefix option to the setClassName helper 1`] = `
<div
  className="tw-a tw-ab"
  color="B"
  id="3"
>
  <div
    className="child"
  >
    Type color classes
  </div>
</div>
`;

exports[`passes the setClassName helper and reacts to prop changes 1`] = `
Array [
  <div
    className=""
    color="D"
    id="1"
  >
    <div
      className="child"
    >
      No classes
    </div>
  </div>,
  <div
    className="a"
    color="D"
    id="2"
  >
    <div
      className="child"
    >
      Type classes
    </div>
  </div>,
  <div
    className="a ab"
    color="B"
    id="3"
  >
    <div
      className="child"
    >
      Type color classes
    </div>
  </div>,
]
`;

exports[`renders the wrapped component as you would expect using the withSetClassName HOC 1`] = `
<div
  className=""
  id="foo"
>
  <div
    className="child"
  >
    Hello!
  </div>
</div>
`;
